---
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: show
    highlight: zenburn
    theme: flatly
---

## Load data

```{r load-data}
infn = "../data/Int Hypox Invitro and Human Meta Data.csv"
metadata = read.table(infn, sep=",", header=TRUE, stringsAsFactors=FALSE)
```

Here we make the coding for the variables easier to work with by making them
smaller and removing spaces and other problem-causing characters. We will
also make an ID that has the important metadata in it, to make plots
easier to interpet.

```{r recode-metdata}
date2month = function(str) {
  month = months(as.Date(str, format="%m.%d.%y"))
  return(substring(month, 1, 3))
}

metadata$month = as.factor(unlist(lapply(metadata$Experiment.Date, date2month)))
metadata$line = as.factor(gsub(" ", "", metadata$Cell.Line))
metadata$spiked = as.factor(ifelse(grepl("No", metadata$Spike.In ), "nospike",
                                   "spike"))
metadata$hypoxia = as.factor(ifelse(grepl("Intermittent", metadata$Condition), "IH",
                             ifelse(grepl("Sustained", metadata$Condition), "SH",
                                    "NO")))
metadata$hypoxia = relevel(metadata$hypoxia, ref="NO")
metadata$filename = metadata$RCC.FILE.NAME
metadata$fullname = metadata$Full.Sample.Name
metadata$column = ifelse(grepl("Oligo", metadata$Column), "oligo",
                  ifelse(grepl("Reg", metadata$Column), "reg", "nocol"))
metadata$replicate = unlist(lapply(metadata$Replicate.Number,
                                   function(x) strsplit(as.character(x), "-")[[1]][2]))
metadata$id = as.character(paste(metadata$line, metadata$hypoxia, metadata$spiked,
                                 metadata$column, metadata$month, sep="_"))
metadata$id = ifelse(grepl("Blank", metadata$id), metadata$Full.Sample.Name,
                     metadata$id)
metadata$linehypo = paste(metadata$line, metadata$hypoxia, sep="_")
```

Now we are ready to roll and can read in the RCC files using the sweet
`NanoStringQCPro` library. We'll just read it in, extract the counts and
then actually do the analysis with DESeq2.

```{r read-nstring}
library(NanoStringQCPro)
rccFiles = paste("../data/rcc/", metadata[,"filename"], sep="")
#blanks = c("20160602_ES052616miRNAV3aC2_Blank_11.RCC", "20160602_ES052616miRNAV3aC2_Blank_12.RCC")
eset = newRccSet(rccFiles=rccFiles, blankLabel="Blank")
pdata = pData(eset)
pdata$SampleType = metadata$line
pData(eset) <- pdata
proc = preprocRccSet(eset)
counts = as.data.frame(exprs(eset))
colnames(counts) = metadata$id
```

```{r predicate-functions}
is_positive = function(column) {
  return(grepl("Pos", column))
}
is_negative = function(column) {
  return(grepl("Neg", column))
}
is_spikein = function(column) {
  return(grepl("Spike", column))
}
is_ligation = function(column) {
  return(grepl("Ligati", column))
}
is_housekeeping = function(column) {
  return(grepl("Housekee", column))
}
is_prior = function(column) {
  return(grepl("miR-210", column) | grepl("miR-365", column))
}
```

## Scale miRNA with known crosstalk
A small set of miRNA have known inflated values; Nanostring has a heuristic to
correct for the inflation via the formula using some supplied constants. Here
we implement that fix to correct the expression of those miRNA.

```{r add-scaledata}
correct_miRNA = function(eset) {
  require(dplyr)
  fdat = fData(eset)
  fdat = fdat %>%
    tidyr::separate(GeneName, c("Name", "scalefactor"), sep='\\|',
                    fill="right", remove=FALSE) %>%
    dplyr::mutate(scalefactor=ifelse(is.na(scalefactor), 0, scalefactor))
  sf = as.numeric(fdat$scalefactor)
  posa = as.numeric(counts["Positive_POS_A_ERCC_00117.1",])
  scales = t(replicate(length(sf), posa))
  scales = scales * sf
  scaled = counts - scales
  scaled[scaled<0] = 0
  rownames(scaled) = fdat$Name
  return(round(scaled))
}

plot_corrected_miRNA = function(scaled, counts) {
  require(reshape)
  require(ggplot2)
  rownames(counts) = rownames(scaled)
  is_scaled = rowSums(abs(counts - scaled)) > 0
  counts = melt(as.matrix(counts[is_scaled,]))
  colnames(counts) = c("gene", "sample", "value")
  counts$scaled = "no"
  scaled = melt(as.matrix(scaled[is_scaled,]))
  colnames(scaled) = c("gene", "sample", "value")
  scaled$scaled = "yes"
  all = rbind(counts, scaled)
  library(cowplot)
  ggplot(all, aes(sample, value, color=scaled)) + facet_wrap(~gene) +
    geom_point(size=0.5) +
    theme(axis.title.x=element_blank(),
          axis.text.x=element_blank(),
          axis.ticks.x=element_blank(),
          strip.text.x=element_text(size=8)) +
    guides(colour = guide_legend(override.aes = list(size=6))) + ylab("") +
    ggtitle("miRNA hybridization crosstalk correction")
}
```

```{r scale-counts}
scaled = correct_miRNA(eset)
plot_corrected_miRNA(scaled, counts)
```

## Imaging QC
This percentage should be super high, close to 100%. If there isn't that usually means
the cartridge needs to be rescanned. If it is < 3 weeks from the scan, rescanning
should be okay. Nanostring folks call < 75% a failure.

```{r imaging-qc}
library(cowplot)
pdat = pData(eset) %>% left_join(metadata, by=c("FileName"="RCC.FILE.NAME"))
pdat$pcounted = pdat$FovCounted/pdat$FovCount * 100
ggplot(pdat, aes(id, pcounted)) + geom_point() +
    theme(axis.text.x=element_blank(),
          axis.ticks.x=element_blank(),
          strip.text.x=element_text(size=8)) +
  scale_y_continuous(expand = c(0,0)) +
  expand_limits(y = c(0,1.05 * max(pdat$pcounted))) +
  ylab("percentage of FOV counted") + xlab("sample") +
  geom_hline(yintercept=75, color="red")
```

## Binding density
Nanostring folks call a sample problematic if the binding density is < 0.05 or
> 2.25.

```{r binding-density}
ggplot(pdat, aes(id, BindingDensity)) + geom_point() +
    theme(axis.text.x=element_blank(),
          axis.ticks.x=element_blank(),
          strip.text.x=element_text(size=8)) +
  scale_y_continuous(expand = c(0,0)) +
  expand_limits(y = c(0,1.05 * max(pdat$BindingDensity))) +
  ylab("Binding density") + xlab("sample") +
  geom_hline(yintercept=0.05, color="red") +
  geom_hline(yintercept=2.25, color="red")
```

## Total counts vs miRNA detected
Here I plotted the total counts vs the number of miRNA detected, where
detected is it has counts > 30. There is a pretty big spread among the samples
in terms of the miRNA detected. The Huh7 cells look like they have a lower
number of genes detected on average for a given number of reads. This
might be a technical artifact or it could also be that there are less miRNA
in the Huh7 cells.

```{r complexity-plot}
endocounts = counts[grepl("Endo", rownames(counts)),]
cdf = data.frame(total=colSums(counts), detected=colSums(counts > 30))
rownames(cdf) = colnames(counts)
cdf$id = rownames(cdf)
cdf = cdf %>% left_join(metadata, by="id")
ggplot(cdf, aes(total, detected, color=column, label=linehypo)) +
  geom_text() + scale_x_log10()
```

## Positive controls
Below we look at the R^2 correlation between the expected positive
control concentrations and the observed concentrations for each
sample.

```{r pos-r2}
pcdf = data.frame(concentration=log2(c(128, 32, 8, 2, 0.5, 0.125)),
                  GeneName=paste("POS", c("A", "B", "C", "D", "E", "F"), sep="_"))
pccounts = subset(exprs(eset), grepl("Positive_POS", rownames(exprs(eset))))
pccounts = pccounts[sort(rownames(pccounts)),]
rownames(pccounts) = pcdf$GeneName
corsamples = data.frame(correlation=apply(pccounts, 2,
                                          function(x) cor(x, pcdf$concentration)),
                        sample=colnames(pccounts)) %>%
  left_join(metadata, by=c("sample"="RCC.FILE.NAME"))
ggplot(corsamples, aes(sample, correlation)) + geom_point() +
    theme(axis.text.x=element_blank(),
          axis.ticks.x=element_blank(),
          strip.text.x=element_text(size=8)) +
  scale_y_continuous(expand = c(0,0)) +
  expand_limits(y = c(0,1.05 * max(corsamples$correlation))) +
  ylab("positive control correlation") +
  xlab("positive control correlation")
```

## Ligation controls
Here we look at R^2 for the ligation controls as well. It looks like a couple
samples failed.

```{r lig-r2}
pcdf = data.frame(concentration=log2(c(128, 32, 8)),
                  GeneName=paste("POS_", c("A", "B", "C"), sep="_"))
pccounts = subset(exprs(eset), grepl("Ligation_LIG_POS", rownames(exprs(eset))))
pccounts = pccounts[sort(rownames(pccounts)),]
rownames(pccounts) = pcdf$GeneName
corsamples = data.frame(correlation=apply(pccounts, 2,
                                          function(x) cor(x, pcdf$concentration)),
                        sample=colnames(pccounts)) %>%
  left_join(metadata, by=c("sample"="RCC.FILE.NAME"))
ggplot(corsamples, aes(sample, correlation)) + geom_point() +
    theme(axis.text.x=element_blank(),
          axis.ticks.x=element_blank(),
          strip.text.x=element_text(size=8)) +
  scale_y_continuous(expand = c(0,0)) +
  expand_limits(y = c(0,1.05 * max(corsamples$correlation))) +
  ylab("ligation control correlation") +
  xlab("ligation control correlation")
```

```{r which-failed-ligation}
subset(corsamples, correlation < 0.75)$sample
```

Ah, of course, the blanks.

```{r positive-controls}
library(ggplot2)
library(dplyr)
library(cowplot)

extract_pred = function(counts, predicate) {
  toplot = counts[predicate(rownames(counts)),] %>%
    tibble::rownames_to_column() %>%
    tidyr::gather("sample", "count", -rowname)
  colnames(toplot) = c("spot", "sample", "count")
  toplot = toplot %>% left_join(metadata, by=c("sample"="id"))
  return(toplot)
}
spotbarplot = function(toplot) {
  ggplot(toplot,
        aes(sample, count)) + geom_bar(stat='identity') +
    facet_wrap(~spot) +
    theme(axis.text.x = element_blank(),
          text = element_text(size=8))
}
spotboxplot = function(toplot) {
  ggplot(toplot,
        aes(linehypo, count)) + geom_boxplot() +
    facet_wrap(~spot) +
    theme(axis.text.x = element_text(angle = 90, hjust = 1))
}
spotbarplot(extract_pred(counts, is_positive))
```

## Negative controls
Negative counts can be relatively high in some of the samples compared to
the other samples.

```{r negative-controls}
spotbarplot(extract_pred(counts, is_negative))
```

```{r high-negative}
knitr::kable(subset(extract_pred(counts, is_negative), count > 50))
```

Was anything different about those samples?

## SpikeIn
It looks like there are only some samples that have spike-ins.
```{r spikein}
spotbarplot(extract_pred(counts, is_spikein))
```

And those correspond to the spike-ins in the metadata:

```{r spike-in-samples}
knitr::kable(subset(extract_pred(counts, is_spikein), count > 500))
```

## Ligation
It looks like some of the ligation controls worked better than the other.
Kit suggested just normalizing by the mean of the posA ligation control.
What we will do instead is do the normalization we usually do, but include
the ligation efficiency for each sample as a factor in the regression.
Then we can test whether or not our normalization scheme is correcting for
the ligation efficiency. It should be, if the result of the ligation efficiency
is overall less miRNA.

```{r ligation}
spotbarplot(extract_pred(counts, is_ligation))
```

### Ligation efficiency

Here we calculated ligation efficiency for the samples. This is the
log of LIG_POS_A(i) / mean(LIG_POS_A) for each sample i. It doesn't matter
which direction this is in, it will get corrected for in the model.

```{r ligation_efficiency}
lig_posa = as.numeric(counts[grepl("Ligation_LIG_POS_A", rownames(counts)),])
metadata$lignorm = log2(lig_posa / mean(lig_posa))
ggplot(subset(metadata, line != "Blank"), aes(id, lignorm)) + geom_point() +
    theme(axis.text.x=element_blank(),
          axis.ticks.x=element_blank(),
          strip.text.x=element_text(size=8)) +
  ylab("ligation factor") +
  xlab("sample")
```

## Housekeeping
For a subset of the samples, one housekeeping gene is very high.
```{r housekeeping}
spotbarplot(extract_pred(counts, is_housekeeping))
```

These don't look confined to any particular cell type. Why are some of these
so different?

```{r housekeeping-samples}
table(subset(extract_pred(counts, is_housekeeping), count > 1000)$sample)
```

## Normalization

## Drop non-negative control and housekeeping
```{r drop-unusable}
drop = is_spikein(rownames(counts))
drop = drop | is_positive(rownames(counts))
drop = drop | is_housekeeping(rownames(counts))
drop = drop | is_ligation(rownames(counts))
keep = counts[!drop,]
keep = keep[, !grepl("Blank", colnames(keep))]
metadata = subset(metadata, id %in% colnames(keep))
counts = keep
```
## Drop genes below noise floor
We'll normalize the libraries and look at the negative control expression and
then come up with a cutoff that is above the noise floor for the libraries.

```{r normalize-deseq2}
library(DESeq2)
dds = DESeqDataSetFromMatrix(counts, colData=metadata, design=~month+column+line+hypoxia+lignorm)
dds = estimateSizeFactors(dds)
ncounts = data.frame(counts(dds, normalized=TRUE))
```

Below we can see after normalizing by library size, if we set the cutoff to be
30, we are mostly above the noise floor measured by the negative controls:

```{r neg-cutoff}
ggplot(extract_pred(ncounts, is_negative), aes(count)) + geom_histogram()
```

We used the threshold of 30 as the noise floor and dropped all" miRNA that did
not have at least 30 counts in 20% of the samples.

```{r drop-mirna}
keepgenes = rowSums(ncounts > 30) > 0.2*ncol(ncounts)
counts = counts[keepgenes,]
```

## Does normalization for ligation effiency buy us anything?
Here we fit two models. The first model we fit the full model,
including the ligation normalization term in the model.
~month+column+line+hypoxia+lignorm. Then we fit a reduced model without
the lignorm term to answer the question 'is there miRNA that are affected by the
ligation efficiency?'

Yes, it looks like even after normalization, if we fit a model with just
the ligation efficiency that there are miRNA affected. So we should correct
for that. Here you can see only a subset of the miRNA are affected; if we
do it like this instead of just scaling the counts, we can correct for
the specific miRNA that have counts that are correlated with the ligation
efficiency.

```{r liga}
dds = DESeqDataSetFromMatrix(counts, colData=metadata, design=~month+column+line+hypoxia+lignorm)
dds = DESeq(dds, full=~month+column+line+hypoxia+lignorm,
            reduced=~month+column+line+hypoxia,  test="LRT")
res = results(dds)
plotMA(res)
res = data.frame(res) %>% tibble::rownames_to_column() %>%
  arrange(padj) %>% filter(padj < 0.05)
knitr::kable(subset(res, padj < 0.05))
```


## PCA
Here we do PCA to look at how the samples cluster. We can see a clear separation
along the 1st and 2nd PCA based on line, but within a cell line, the different
hypoxic states overlap.

```{r de-setup}
dds = DESeqDataSetFromMatrix(counts, colData=metadata,
                             design=~column+spiked+line+hypoxia+lignorm)
vst = varianceStabilizingTransformation(dds)
pca_loadings = function(object, ntop=500) {
  rv <- matrixStats::rowVars(assay(object))
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop,
      length(rv)))]
  pca <- prcomp(t(assay(object)[select, ]))
  percentVar <- pca$sdev^2/sum(pca$sdev^2)
  names(percentVar) = colnames(pca$x)
  pca$percentVar = percentVar
  return(pca)}
pc = pca_loadings(vst)
comps = data.frame(pc$x)
comps$Name = rownames(comps)
library(dplyr)
comps = comps %>% left_join(metadata, by=c("Name"="id"))
pca_plot = function(comps, nc1, nc2, colorby) {
   c1str = paste0("PC", nc1)
   c2str = paste0("PC", nc2)
  ggplot(comps, aes_string(c1str, c2str, color=colorby)) +
    geom_point() + theme_bw() +
    xlab(paste0(c1str, ": ", round(pc$percentVar[nc1] * 100), "% variance")) +
    ylab(paste0(c2str, ": ", round(pc$percentVar[nc2] * 100), "% variance"))
  }
pca_plot(comps, 1, 2, "linehypo")
pca_plot(comps, 3, 4, "linehypo")
```

## Differential expression

Nanostring data has an opposite mean-variance relationship than RNA-seq data.
Here the dispersion is slightly higher for more highly expressed miRNA-- the
the trend is mostly flat, however.

```{r mirna-dispersion}
dds = DESeq(dds, fitType='local')
plotDispEsts(dds)
```

## Cell line differences
Here we ask, controlling for the month, the column and the hypoxic state of the
cell line, are there differences between the cell lines. The cell lines are
very different from each other, most of the miRNA are significantly different
between the cell lines.

### Huh7 vs U937
```{r huh7-vs-u937}
res = results(dds, contrast=c("line", "Huh7", "U937"))
```

### Huh7 vs THP
```{r huh7-vs-thp}
res = results(dds, contrast=c("line", "Huh7", "THP"))
```

### THP vs U937
```{r THP-vs-U937}
res = results(dds, contrast=c("line", "THP", "U937"))
```

## Hypoxic state differences
Here we are asking, controlling for the month, column and the cell line, are
there differences due to hypoxic state?

### Normal vs intermittent
```{r no-vs-ih}
res = results(dds, contrast=c("hypoxia", "NO", "IH"))
```

### Intermittent vs sustained
```{r ih-vs-sh}
res = results(dds, contrast=c("hypoxia", "IH", "SH"))
```

### Normal vs sustained
```{r no-vs-sh}
res = results(dds, contrast=c("hypoxia", "NO", "SH"))
```

## Cell line specific hypoxic state differences
Here we refit the model with a new factor that is a combination of cell line
and hypoxic state. Then we test if there are differences between the hypoxic
states within each cell line.

```{r refit-model}
metadata$linehypo = paste(metadata$line, metadata$hypoxia, sep="_")
dds = DESeqDataSetFromMatrix(counts, colData=metadata,
                             design=~column+spiked+linehypo+lignorm)
dds = DESeq(dds, fitType='local')
```

### THP
We can se that the overall fold changes are fairly small relative to the
standard error of the fold changes. Signal to noise ratio here is combined
as the log fold change divided by the log fold change standard error.

```{r THP}
write_results = function(res, fname) {
  res = data.frame(res) %>% tibble::rownames_to_column() %>%
    arrange(pvalue)
  write.table(res, file=fname, quote=FALSE, col.names=TRUE,
              row.names=FALSE, sep=",")
}

res = results(dds, contrast=c("linehypo", "THP_NO", "THP_IH"), addMLE=TRUE)
ggplot(data.frame(res), aes(log2FoldChange, lfcMLE/lfcSE)) + geom_point() +
  ylab("signal to noise ratio") + ggtitle("THP NO vs IH")
write_results(res, "THP_NOvsIH.csv")
res = results(dds, contrast=c("linehypo", "THP_NO", "THP_SH"), addMLE=TRUE)
ggplot(data.frame(res), aes(log2FoldChange, lfcMLE/lfcSE)) + geom_point() +
  ylab("signal to noise ratio") + ggtitle("THP NO vs SH")
write_results(res, "THP_NOvsSH.csv")
res = results(dds, contrast=c("linehypo", "THP_IH", "THP_SH"), addMLE=TRUE)
ggplot(data.frame(res), aes(log2FoldChange, lfcMLE/lfcSE)) + geom_point() +
  ylab("signal to noise ratio") + ggtitle("THP IH vs SH")
write_results(res, "THP_IHvsSH.csv")
```

### Huh7
```{r Huh7}
res = results(dds, contrast=c("linehypo", "Huh7_NO", "Huh7_IH"), addMLE=TRUE)
ggplot(data.frame(res), aes(log2FoldChange, lfcMLE/lfcSE)) + geom_point() +
  ylab("signal to noise ratio") + ggtitle("Huh7 NO vs IH")
write_results(res, "Huh7_NOvsIH.csv")
res = results(dds, contrast=c("linehypo", "Huh7_NO", "Huh7_SH"), addMLE=TRUE)
ggplot(data.frame(res), aes(log2FoldChange, lfcMLE/lfcSE)) + geom_point() +
  ylab("signal to noise ratio") + ggtitle("Huh7 NO vs SH")
write_results(res, "Huh7_NOvsSH.csv")
res = results(dds, contrast=c("linehypo", "Huh7_IH", "Huh7_SH"), addMLE=TRUE)
ggplot(data.frame(res), aes(log2FoldChange, lfcMLE/lfcSE)) + geom_point() +
  ylab("signal to noise ratio") + ggtitle("Huh7 IH vs SH")
write_results(res, "Huh7_IHvsSH.csv")
```

### U937

```{r u937}
res = results(dds, contrast=c("linehypo", "U937_NO", "U937_IH"), addMLE=TRUE)
ggplot(data.frame(res), aes(log2FoldChange, lfcMLE/lfcSE)) + geom_point() +
  ylab("signal to noise ratio") + ggtitle("U937 NO vs IH")
write_results(res, "U937_NOvsIH.csv")
res = results(dds, contrast=c("linehypo", "U937_NO", "U937_SH"), addMLE=TRUE)
ggplot(data.frame(res), aes(log2FoldChange, lfcMLE/lfcSE)) + geom_point() +
  ylab("signal to noise ratio") + ggtitle("U937 NO vs SH")
write_results(res, "U937_NOvsSH.csv")
res = results(dds, contrast=c("linehypo", "U937_IH", "U937_SH"), addMLE=TRUE)
ggplot(data.frame(res), aes(log2FoldChange, lfcMLE/lfcSE)) + geom_point() +
  ylab("signal to noise ratio") + ggtitle("U937 IH vs SH")
write_results(res, "U937_IHvsSH.csv")
```

## Prior miRNA
miR-210 is below the noise floor-- one of the miR-365 spots though has
higher expression values.
```{r plot-prior}
spotbarplot(extract_pred(counts, is_prior)) + scale_y_log10()
spotboxplot(extract_pred(counts, is_prior))
```

## Ouput tables
[Huh7 NO vs IH](Huh7_NOvsIH.csv)

[Huh7 SO vs IH](Huh7_SOvsIH.csv)

[Huh7 IH vs SH](Huh7_IHvsSH.csv)

[THP NO vs IH](THP_NOvsIH.csv)

[THP SO vs IH](THP_SOvsIH.csv)

[THP IH vs SH](THP_IHvsSH.csv)

[U937 NO vs IH](U937_NOvsIH.csv)

[U937 SO vs IH](U937_SOvsIH.csv)

[U937 IH vs SH](U937_IHvsSH.csv)
